### ISSUE TITLE
Improved Handling of Shard Loss Scenario

### ISSUE THREAD

**Issue by danielyahn** [elastic/elasticsearch-hadoop#1337]
### What kind an issue is this?
- [x] Feature Request. 
 
### Issue description
We use ES-Hadoop library in batch spark jobs which are scheduled every quarter hour to ingest data into ElasticSearch. The jobs are continually scheduled so that ingestion stream keeps coming in.

We've seen several cases where the cluster goes Red and then ingestion completely stops. Here is my understanding of this failure mode:

1. ES-Hadoop job sends a bulk request `t`, write to primary shard `p1` of index `i_20190923-13` is successful, but write to replica `r1.1` and `r1.2` haven't happened yet
2. ES detects the hardware issue, where `p1` resides. And it tries to recover `p1` from `r1.1` and `r1.2`. ES can't find `r1.1` and `r.1.2` because replica writes haven't completed. Cluster marks this index index `i_20190923-13` as red.
3. ES-Hadoop job sends next bulk request `t+1`, but cannot continue, because `p1` got lost.
4. ES-Hadoop job hangs, until we execute reroute API call to `p1` with `accept_data_loss=true`
5. No more Spark can get scheduled due to limited compute resource because these hung ES-Hadoop jobs are holding onto them

### Feature description
So the key theme of my query to continue ingestion even when cluster is red.

Few ideas:
1. Set timeouts on Spark job, so next job in schedule gets started and creates new hourly indices (`i_20190923-14`) on a good hardware. We can make this as either built-in to ES-Hadoop library or as an outside watcher of Spark job.
2. Get acknowledge of replica write on bulk API calls to prevent problems of writes to active primary shard gets lost. There will be performance degradation.
3. ES-Hadoop job keeps track of bulk API calls that haven't completed replica writes. So that in this scenario when we lose primary shard `p1` and have no replica `r1.1` and `r1.2`, it creates a new empty primary shard `p1` on a good hardware, then replay that bulk API  request `t`.

**Comment by jbaiera** [comment]
> ES-Hadoop job hangs, until we execute reroute API call to p1 with accept_data_loss=true

This seems like a considerable bug. While the job is hanging, can you include a thread dump to help us debug this? We definitely should ensure that cluster resources on Hadoop/Spark aren't blocked up waiting on a failing write operation.

> Few ideas:

I agree that we should always be looking for ways to improve our handling of failure scenarios, but a key thing to remember here is that these sorts of changes have ripple effects. Assume that ES adds the ability you suggest of waiting on at least some replication to complete before returning. Not only is this a performance degradation due to response time from the cluster for the original bulk request, but it is also now adding the expectation that ES-Hadoop does something with this information. As you suggest, we could hold on to the bulk request and retry it at a later time, but bulk requests from ES-Hadoop could be fairly large in size, and keeping them around to retry them later would be expensive.

Additionally, having ES-Hadoop perform administrative operations like allocating missing shards directly feels like the wrong direction here. `accept_data_loss` is one of those options that is meant to only be performed by an administrator of the system after very careful thought and understanding the ramifications of the action. Generally, Elasticsearch can be trusted to perform the recovery of missing shards on its own because it is aware of the full cluster state. If Elasticsearch does not see a way to perform a successful recovery of a missing primary (because of missing writes in the replica), then it is unlikely that ES-Hadoop would be able to do better, even if it has managed to keep track of write operations that it sees were not replicated.

The ES-Hadoop project has followed the mantra of retry where possible, allow for intervention where reasonable, but fail fast otherwise. I'd love to hear your thoughts on the matter (and to hopefully get some more info about the hanging jobs because that's no good).

**Comment by danielyahn** [comment]
Yes, I agree that spark job should fail fast. This failure is some what challenging to reproduce because it was caused by hard drive issue. I'll share the information as I gather more.

**Comment by jbaiera** [comment]
@danielyahn since there isn't much that is actionable in this issue, I'm going to go ahead and close it out. If you are able to send along any information about ES-Hadoop becoming unresponsive when trying to read from a red cluster, please open a new issue on here with that info and reference this issue in the new one for context. Thanks!
